src/main/java/io/micronaut/build/DockerCracMojo.java

```diff
@@ -0,0 +1,206 @@
+                    "https://docs.docker.com/engine/reference/commandline/tag/#extended-description" +
+                    "\nFor example, slash-separated name components cannot have uppercase letters";
+            throw new MojoExecutionException(message);
+        } catch (IOException | IllegalArgumentException e) {
+            throw new MojoExecutionException(e.getMessage(), e);
+        }
+    }
+
+    private void buildDockerCrac() throws IOException, InvalidImageReferenceException {
+        String checkpointImage = buildCheckpointDockerfile();
+        getLog().info("CRaC Checkpoint image: " + checkpointImage);
+        dockerService.runPrivilegedImageAndWait(checkpointImage, new File(mavenProject.getBuild().getDirectory(), "cr").getAbsolutePath() + ":/home/app/cr");
+        buildFinalDockerfile(checkpointImage);
+    }
+
+    private String buildCheckpointDockerfile() throws IOException {
+        String name = mavenProject.getArtifactId() + "-crac-checkpoint";
+        Set<String> checkpointTags = Collections.singleton(name);
+        copyScripts(CHECKPOINT_SCRIPT_NAME, WARMUP_SCRIPT_NAME, RUN_SCRIPT_NAME);
+        File dockerfile = dockerService.loadDockerfileAsResource(DockerfileMojo.DOCKERFILE_CRAC_CHECKPOINT);
+        BuildImageCmd buildImageCmd = dockerService.buildImageCmd()
+                .withDockerfile(dockerfile)
+                .withBuildArg("BASE_IMAGE", getFromImage().orElse(DEFAULT_BASE_IMAGE))
+                .withTags(checkpointTags);
+        dockerService.buildImage(buildImageCmd);
+        return name;
+    }
+
+    private String getReadinessCommand() {
+        String property = mavenProject.getProperties().getProperty(DockerCracMojo.CRAC_READINESS_PROPERTY);
+        return property == null ? DockerCracMojo.DEFAULT_READINESS_COMMAND : property;
+    }
+
+    private void buildFinalDockerfile(String checkpointContainerId) throws IOException, InvalidImageReferenceException {
+        Set<String> tags = getTags();
+        for (String tag : tags) {
+            ImageReference.parse(tag);
+        }
+        copyScripts(RUN_SCRIPT_NAME);
+        File dockerfile = dockerService.loadDockerfileAsResource(DockerfileMojo.DOCKERFILE_CRAC);
+        BuildImageCmd buildImageCmd = dockerService.buildImageCmd()
+                .withDockerfile(dockerfile)
+                .withBuildArg("BASE_IMAGE", getFromImage().orElse(DEFAULT_BASE_IMAGE))
+                .withBuildArg("CHECKPOINT_IMAGE", checkpointContainerId)
+                .withTags(getTags());
+        dockerService.buildImage(buildImageCmd);
+    }
+
+    private void copyScripts(String... scriptNames) throws IOException {
+        List<String> imageClasspathScopes = Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+        String readinessCommand = getReadinessCommand();
+        mavenProject.setArtifactFilter(artifact -> imageClasspathScopes.contains(artifact.getScope()));
```

Comment:
Why is this needed?

Reply:
I copied it from somewhere...  My lack of Maven skills is showing 😉 Gone now